### PR TITLE
Rewrite player switching in multiplayer dialog

### LINKED ISSUE #2378: Cannot switch players in multiplayer mode
After we choose one player, we cannot switch with another human player.
image

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
fix #2378

Please note that this PR also implements an ability to swap player names & classes in multiplayer mode even if players aren't supposed to be humans. For example, we cannot swap GOOD and BAD players here:

Switch failed

But we are able to swap BAD and UGLY:

Switch OK

I also changed the logic of the swap buttons on the bottom of the dialog window. Previously, they used to swap absolutely everything, including color (?!). Now they work exactly the same way as using mouse. If this is an unwanted change, I can roll it back.

### CODE REVIEW

File: src/fheroes2/gui/player_info.cpp [hunk 1]
```diff
@@ -72,6 +72,66 @@ void Interface::PlayersInfo::UpdateInfo( Players & players, const Point & pt1, c
...
 
+bool Interface::PlayersInfo::SwapPlayers( Player & player1, Player & player2 ) const
+{
+    const Settings & conf = Settings::Get();
+    const Maps::FileInfo & fi = conf.CurrentFileInfo();
+
+    const int player1Color = player1.GetColor();
+    const int player2Color = player2.GetColor();
+
+    bool swap = false;
```
Comment (ihhub) [inline src/fheroes2/gui/player_info.cpp:83]: We could remove this variable by doing this:
```
if ( ( player1Color & fi.AllowCompHumanColors() ) && ( player2Color & fi.AllowCompHumanColors() ) ) {
    // the code
}
else if ( player1.isControlAI() != player2.isControlAI() ) {
    return false;
}

// the rest of the code

return true;
```

**Review by ihhub — COMMENTED** [review]
Hi @oleg-derevenetz , the code looks good. I left just one optional comment. I'm leaving the decision for you.

**Comment by ihhub** [comment]
Hi @Branikolog , could you please test this pull request?

File: src/fheroes2/gui/player_info.cpp (hunk 1 above)
Comment (oleg-derevenetz) [reply src/fheroes2/gui/player_info.cpp:83]: Hi @ihhub 

> We could remove this variable by doing this:

I think that's not entirely equivalent code. The thing is, we shouldn't mess with SetPlayerControl() at all if both players are already humans or both players are already computers, so we should check this first. AllowCompHumanColors() tells us just a POSSIBILITY for a player to be either human or computer, it says nothing about who currently controls the player, so we will need some additional checks somewhere inside the first if anyway. I think this is an unnecessary complication.

**Comment by Branikolog** [comment]
Everything seems to work fine!
Actually, I'm not sure about switching arrows... Why do we need them here? But right now they work predictable. Let's leave them for now as @oleg-derevenetz implemented.

**Comment by oleg-derevenetz** [comment]
Hi @Branikolog

> Actually, I'm not sure about switching arrows... Why do we need them here?

That's an interesting question, because all they did before was change the order of the colors itself. I don't know why it was needed.

**Comment by Branikolog** [comment]
@oleg-derevenetz 
Probably, previous team wanted to implement changing the order of players in game? It's the only purpose I could imagine for these buttons...

**Comment by ihhub** [comment]
@oleg-derevenetz , great work! Thank you for these changes!

**Pull request merged**